File: apps/dotnet/Morphir.Cli/CommandLineHost.fs
namespace Morphir.Cli

open System
open Spectre.Console.Cli
open Morphir.Cli.Commands

type HostConfig = {
    ApplicationName: string
}

type CommandLineHost(hostConfig:HostConfig) = 
    let app = CommandApp()
    do app.Configure(fun config ->
        config.SetApplicationName(hostConfig.ApplicationName) |> ignore
        config.AddCommand<Build.Build>("build")
            .WithDescription("Builds the project.")         
            |> ignore
        config.AddCommand<Run.Run>("run")
            .WithDescription("Runs a model.")         
            |> ignore
        
        config.AddBranch<Dependency.DependencySettings>("dependency", fun dependency ->
            dependency.AddCommand<Dependency.Refresh>("refresh")
                .WithDescription("Refreshes dependencies.")         
                |> ignore
            ()
        ) |> ignore
    )                                                             
        
    member this.Run(argv) =
        app.Run(argv)

module CommandLineHost =
    let runHost (argv:string seq) (config:HostConfig) =
        let host = CommandLineHost(config)
        host.Run argv
                 
    let inline run(argv: string seq, config:HostConfig) =
        runHost argv config

File: apps/dotnet/Morphir.Cli/CommandLineHost.cs
using System.Collections.Generic;
using Morphir.Cli.Commands.Dependency;
using Morphir.Host;
using Spectre.Console.Cli;

namespace Morphir.Cli;

public class CommandLineHost(HostConfig hostConfig)
{
    private CommandApp _app = CreateApp(hostConfig);

    private static CommandApp CreateApp(HostConfig hostConfig)
    {
        var app = new CommandApp();
        app.Configure(config =>
        {
            config.AddCommand<Commands.Build.BuildCommand>("build")
                .WithDescription("Build the project");
            config.AddCommand<Commands.Run.RunCommand>("run")
                .WithDescription("Run the project");
            config.AddBranch<DependencySettings>("dependency", dependencyConfig =>
            {
                dependencyConfig.AddCommand<DependencyFetchCommand>("fetch")
                    .WithDescription("Fetch dependencies");
            });
        });
        return app;
    }
    
    public int Run(IEnumerable<string> args) => _app.Run(args);
    
    public static int Run(HostConfig hostConfig, IEnumerable<string> args) => new CommandLineHost(hostConfig).Run(args);
}